### PR TITLE
Replace `array_contains` with SQL array functions: `array_has`, `array_has_any`, `array_has_all`

### LINKED ISSUE #6973: Implement `array_has_all` function
### Is your feature request related to a problem or challenge?

## Summary

| Characteristic        | Description                                                                                                                                                                                                               |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Function name:        | `array_has_all`                                                                                                                                                                                                           |
| Aliases:              | `list_has_all`, `@>`, `<@`                                                                                                                                                                                                |
| Original function?:   | No                                                                                                                                                                                                                        |
| Function Description: | <b>Clickhouse:</b> Checks whether one array is a subset of another. <br> <b>DuckDB:</b> Returns true if all elements of sub-list exist in list. <b> </b>                                                                  |
| Sources:              | [Concept](https://github.com/apache/arrow-datafusion/discussions/6855) [ClickHouse](https://clickhouse.com/docs/en/sql-reference/functions/array-functions#hasall) [DuckDB](https://duckdb.org/docs/sql/functions/nested) |

## Examples:

```
select array_has_all([1, 2, 3, 4], [3, 4]);
----
true
```

```
select array_has_all([1, 2, 3, 4], [5, 4]);
----
false
```


### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context
https://github.com/apache/arrow-datafusion/issues/6972
_No response_

### LINKED ISSUE #6976: Implement `array_has_any` function
### Is your feature request related to a problem or challenge?

## Summary

| Characteristic        | Description                                                                                                                                                                                                                                                                                                                                                                                                   |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Function name:        | `array_has_any`                                                                                                                                                                                                                                                                                                                                                                                               |
| Aliases:              | `list_has_any`                                                                                                                                                                                                                                                                                                                                                                                          |
| Original function?:   | No                                                                                                                                                                                                                                                                                                                                                                                                            |
| Function Description: | <b>PostgreSQL:</b> Do the arrays overlap, that is, have any elements in common? </br> <b>Azure DataBricks:</b> Returns true if the intersection of array1 and array2 is not empty. </br> <b>Clickhouse:</b> Checks whether two arrays have intersection by some elements. </br> <b>DuckDB:</b> Returns true if any elements exist is both lists. <b> </b>                                                     |
| Sources:              | [Concept](https://github.com/apache/arrow-datafusion/discussions/6855) [PostgreSQL](https://www.postgresql.org/docs/current/functions-array.html) [Azure](https://learn.microsoft.com/en-us/azure/databricks/sql/language-manual/functions/arrays_overlap) [ClickHouse](https://clickhouse.com/docs/en/sql-reference/functions/array-functions#hasany) [DuckDB](https://duckdb.org/docs/sql/functions/nested) |

## Examples:

```
select has_any([1, 2, 3, 4], [3, 5, 6]);
----
true
```

```
select has_any([1, 5, 3, 4], [2, 6]);
----
false
```


### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #6977: Implement `array_has` function
### Is your feature request related to a problem or challenge?

## Summary

| Characteristic        | Description                                                                                                                                                                                                                                                                                                                             |
| --------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Function name:        | `array_has`                                                                                                                                                                                                                                                                                                                             |
| Aliases:              | `list_has`, `array_contains`, `list_contains`                                                                                                                                                                                                                                                                                           |
| Original function?:   | No                                                                                                                                                                                                                                                                                                                                      |
| Function Description: | <b>Azure DataBricks:</b> Returns true if array contains value. </br> <b>Clickhouse:</b> Checks whether the ‘arr’ array has the ‘elem’ element. Returns 0 if the element is not in the array, or 1 if it is. </br> <b>DuckDB:</b> Returns true if the list contains the element. <b> </b>                                                |
| Sources:              | [Concept](https://github.com/apache/arrow-datafusion/discussions/6855) [Azure](https://learn.microsoft.com/en-us/azure/databricks/sql/language-manual/functions/array_contains) [ClickHouse](https://clickhouse.com/docs/en/sql-reference/functions/array-functions#hasarr-elem) [DuckDB](https://duckdb.org/docs/sql/functions/nested) |

## Examples:

```
select array_has([1, 2, 3, 4], 3);
----
true
```

```
select array_has([1, 2, 3, 4], 5);
----
false
```


### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayzhan211**
# Which issue does this PR close?

Ref #6980 

Close #6973
Close #6976 
Close #6977 

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Review by jayzhan211 — COMMENTED** [review]
Not sure why I got these three files after regen.sh

proto_descriptor.bin
src/datafusion.rs serde.rs

File: datafusion/physical-expr/src/array_expressions.rs [hunk 1]
```diff
@@ -1363,69 +1363,278 @@ pub fn array_ndims(args: &[ArrayRef]) -> Result<ArrayRef> {
...
-/// Array_contains SQL function
-pub fn array_contains(args: &[ArrayRef]) -> Result<ArrayRef> {
-    fn concat_inner_lists(arg: ArrayRef) -> Result<ArrayRef> {
-        match arg.data_type() {
-            DataType::List(field) => match field.data_type() {
-                DataType::List(..) => {
-                    concat_inner_lists(array_concat(&[as_list_array(&arg)?
-                        .values()
-                        .clone()])?)
+/// Array_has_any SQL function
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/array_expressions.rs:1471]: Did not find a way that combines `has_any` and `has_all` without losing readability.

File: datafusion/physical-expr/src/array_expressions.rs [hunk 2]
```diff
@@ -2267,63 +2267,6 @@ mod tests {
         assert_eq!(result, &UInt64Array::from_value(2, 1));
     }
 
-    #[test]
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/array_expressions.rs:2061]: Remove this since `array_has_all` logic changes, and test included in sqllogic

**Comment by jayzhan211** [comment]
@izveigor @alamb Ready for review, thanks

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 3]
```diff
@@ -1009,48 +1063,117 @@ NULL 1 1
 2 NULL 1
 2 1 NULL
 
-## array_contains
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/array.slt:1012]: This looks to me like the tests for `array_contains` have been removed. Was that by accident?

**Review by alamb — COMMENTED** [review]
Should we add `list_has`, `list_has_any`, `list_has_all` aliases to be consistent with https://github.com/apache/arrow-datafusion/pull/7008 from @izveigor ?

For anyone else following along `array_contains` was introduced by @izveigor  in https://github.com/apache/arrow-datafusion/pull/6618 so I don't think it has been released yet (and thus this isn't an API change)

File: datafusion/physical-expr/src/array_expressions.rs [hunk 4]
```diff
@@ -1363,69 +1363,278 @@ pub fn array_ndims(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+        let arr = downcast_arg!($ARRAY, $ARRAY_TYPE);
+        let sub_arr = downcast_arg!($SUB_ARRAY, $ARRAY_TYPE);
+
+        let mut res = false;
+        for elem in sub_arr.iter().dedup() {
+            res |= arr
+                .iter()
+                .dedup()
+                .flatten()
+                .any(|x| x == elem.expect("null type not supported"));
```
Comment (alamb) [inline datafusion/physical-expr/src/array_expressions.rs:1465]: Again, can we please try and avoid panic's on not supported errors -- it makes for a much easier to understand error message for users

File: datafusion/physical-expr/src/array_expressions.rs (hunk 2 above)
Comment (alamb) [reply datafusion/physical-expr/src/array_expressions.rs:2061]: > Sqllogictests are more important than unit tests: unit tests are mainly for convenience.

I agree with this

**Review by alamb — COMMENTED** [review]
Thank you @jayzhan211 

I think this PR looks good to me other than the panic's instead of errors. Once those are fixed I think this one is good to go.

Also, thank you to both you and @izveigor  for pushing these functions ahead -- it is great to see DataFusion grow so nicely

File: datafusion/core/tests/sqllogictests/test_files/array.slt (hunk 3 above)
Comment (jayzhan211) [reply datafusion/core/tests/sqllogictests/test_files/array.slt:1012]: No, I `array_contains` is deprecated, the logic is not the same as `array_has_all` or `array_has`.

File: docs/source/user-guide/sql/scalar_functions.md [hunk 5]
```diff
@@ -1522,19 +1522,49 @@ array_concat(array[, ..., array_n])
...
+#### Arguments
+
+- **array**: Array expression.
+  Can be a constant, column, or function, and any combination of array operators.
+- **sub-array**: Array expression.
+  Can be a constant, column, or function, and any combination of array operators.
+
+### `array_has_any`
+
+Returns true if any elements exist is both array
```
Comment (Weijun-H) [inline docs/source/user-guide/sql/scalar_functions.md:1557]:
```suggestion
Returns true if any elements exist in both arrays
```

File: docs/source/user-guide/expressions.md [hunk 6]
```diff
@@ -179,24 +179,26 @@ Unlike to some databases the math functions in Datafusion works the same way as
...
+| Function                             | Notes                                                                                                          |
+| ------------------------------------ | -------------------------------------------------------------------------------------------------------------- |
+| array_append(array, element)         | Appends an element to the end of an array. `array_append([1, 2, 3], 4) -> [1, 2, 3, 4]`                        |
+| array_concat(array[, ..., array_n])  | Concatenates arrays. `array_concat([1, 2, 3], [4, 5, 6]) -> [1, 2, 3, 4, 5, 6]`                                |
+| array_has(array, element)            | Returns true if the array contains the element `array_has([1,2,3], 1) -> true`                                 |
+| array_has_all(array, sub-array)      | Returns true if all elements of sub-array exist in array `array_has_all([1,2,3], [1,3]) -> true`               |
+| array_has_any(array, sub-array)      | Returns true if any elements exist is both array `array_has_any([1,2,3], [1,4]) -> true`                       |
```
Comment (Weijun-H) [inline docs/source/user-guide/expressions.md:188]:
```suggestion
| array_has_any(array, sub-array)      | Returns true if any elements exist in both arrays `array_has_any([1,2,3], [1,4]) -> true`                       |
```

**Review by alamb — APPROVED** [review]
Thank you @jayzhan211  and @izveigor

**Pull request merged**